WIP: Try to parallelize linting with amphp

staabm opened the pull request:
trying to work on some parallelization, so linting could happen on several CPUs

staabm commented:
@kelunik @trowski could someone of you give me a hand on how this could work.

it seems to work more or less, but I am not sure it works in parallel as it should be.
also I am seeing the following errors on shutdown:

```
$ time php php-cs-fixer fix ../redaxo/redaxo/src/core/
Loaded config default from "C:\xampp7.1.4\htdocs\PHP-CS-Fixer\.php_cs.dist".
Paths from configuration file have been overridden by paths provided as command arguments.
string(30) "PhpCsFixer\Linter\PooledLinter"

Fixed all files in 35.574 seconds, 16.000 MB memory used
PHP Fatal error:  Uncaught ErrorException: proc_terminate(): supplied resource is not a valid process resource in C:\xampp7.1.4\htdocs\PHP-CS-Fixer\vendor\amphp\process\lib\Internal\Windows\Runner.php:154
Stack trace:
#0 [internal function]: {closure}(2, 'proc_terminate(...', 'C:\\xampp7.1.4\\h...', 154, Array)
#1 C:\xampp7.1.4\htdocs\PHP-CS-Fixer\vendor\amphp\process\lib\Internal\Windows\Runner.php(154): proc_terminate(Resource id #634)
#2 C:\xampp7.1.4\htdocs\PHP-CS-Fixer\vendor\amphp\process\lib\Process.php(141): Amp\Process\Internal\Windows\Runner->kill(Object(Amp\Process\Internal\Windows\Handle))
#3 C:\xampp7.1.4\htdocs\PHP-CS-Fixer\vendor\amphp\parallel\lib\Context\Process.php(368): Amp\Process\Process->kill()
#4 C:\xampp7.1.4\htdocs\PHP-CS-Fixer\vendor\amphp\parallel\lib\Worker\Internal\WorkerProcess.php(57): Amp\Parallel\Context\Process->kill()
#5 C:\xampp7.1.4\htdocs\PHP-CS-Fixer\vendor\amphp\parallel\lib\Worker\TaskWorker.php(60): Amp\Parallel\Worker\Internal\WorkerProcess->kill()
#6 [internal function]: Amp\Parallel in C:\xampp7.1.4\htdocs\PHP-CS-Fixer\vendor\amphp\process\lib\Internal\Windows\Runner.php on line 154

Fatal error: Uncaught ErrorException: proc_terminate(): supplied resource is not a valid process resource in C:\xampp7.1.4\htdocs\PHP-CS-Fixer\vendor\amphp\process\lib\Internal\Windows\Runner.php:154
Stack trace:
#0 [internal function]: {closure}(2, 'proc_terminate(...', 'C:\\xampp7.1.4\\h...', 154, Array)
#1 C:\xampp7.1.4\htdocs\PHP-CS-Fixer\vendor\amphp\process\lib\Internal\Windows\Runner.php(154): proc_terminate(Resource id #634)
#2 C:\xampp7.1.4\htdocs\PHP-CS-Fixer\vendor\amphp\process\lib\Process.php(141): Amp\Process\Internal\Windows\Runner->kill(Object(Amp\Process\Internal\Windows\Handle))
#3 C:\xampp7.1.4\htdocs\PHP-CS-Fixer\vendor\amphp\parallel\lib\Context\Process.php(368): Amp\Process\Process->kill()
#4 C:\xampp7.1.4\htdocs\PHP-CS-Fixer\vendor\amphp\parallel\lib\Worker\Internal\WorkerProcess.php(57): Amp\Parallel\Context\Process->kill()
#5 C:\xampp7.1.4\htdocs\PHP-CS-Fixer\vendor\amphp\parallel\lib\Worker\TaskWorker.php(60): Amp\Parallel\Worker\Internal\WorkerProcess->kill()
#6 [internal function]: Amp\Parallel in C:\xampp7.1.4\htdocs\PHP-CS-Fixer\vendor\amphp\process\lib\Internal\Windows\Runner.php on line 154

real    0m44.422s
user    0m0.015s
sys     0m0.030s
```

do you have any hints for me?

staabm commented on composer.json:
```diff
@@ -14,9 +14,10 @@
     ],
     "require": {
-        "php": "^5.6 || ^7.0",
+        "php": "^7.0",
```
will work on how the compat with old php can/should look like, after I get it working on current php versions

julienfalque replied:
You might want to see #4810.

GrahamCampbell replied:
If we don't want PHP 5, then why do we even need a parallel linter? We can use the tokenising linter on PHP 7?

staabm replied:
> why do we even need a parallel linter? We can use the tokenising linter on PHP 7?

I had the theory that tokenized linting in parallel process is faster then doing it in a single process

staabm commented:
it looks like the above reported errors are gone.

it seems the linter does no longer report errors, when invalid files are checked though.

a collegue told me he is getting a "too many open files" error when running the fixer on macOS.

staabm commented:
on the current state of this PR I am getting

![image](https://user-images.githubusercontent.com/120441/75113850-9035b500-5651-11ea-80ee-fb1448a34879.png)

trowski commented on src/Linter/PooledLinter.php:
```diff
@@ -0,0 +1,65 @@
+
+    public function __construct()
+    {
+        $this->processPool = new DefaultPool();
```
I would allow a `Amp\Parallel\Worker\Pool` instance to be passed via the constructor. If one isn't passed, use the one returned from `Amp\Parallel\Worker\pool()`.

trowski commented:
> a collegue told me he is getting a "too many open files" error when running the fixer on macOS.

This can happen if `ulimit` is set too low for the number of processes and open files being created. The default on macOS is a bit low and can be increased using `ulimit -n 8192`, for example. However, since this should be portable, I would reduce the number of workers the pool creates. The default is 32, try reducing it to 16 or 8.

> on the current state of this PR I am getting [image]

On a quick glance, I'm assuming some implementation of `LintingResultInterface` is returning a result that contains a `Closure` that cannot be serialized. Looks like you may need to inspect what is returned and convert it to another implementor of `LintingResultInterface` that can be serialized.

> it seems the linter does no longer report errors, when invalid files are checked though.

Seems the errors are reported as exceptions that likely cannot be serialized, such as `TokenizerLintingResult`. Another `LintingResultInterface` implementation that reports it in a serializable way should take care of that, shouldn't be too complicated. Look at `ContextPanicError` in the current release of `amphp/parallel` for an example of how exception serialization is handled in parallel.

julienfalque commented:
Just wondering, wouldn't it be easier to make PHP CS Fixer more friendly with multiple parallel instances so that it can be used with e.g. [GNU Parallel](https://www.gnu.org/software/parallel/)? Maybe it already is but I never tried it myself.

staabm commented on src/Linter/PooledLintFileTask.php:
```diff
@@ -0,0 +1,24 @@
+     */
+    public function run(Environment $environment) {
+        $linter = new TokenizerLinter();
+        return $linter->lintFile($this->path);
```
```suggestion
        return $linter->lintFile($this->path)->check();
```

staabm commented on src/Linter/PooledLintSourceTask.php:
```diff
@@ -0,0 +1,25 @@
+     */
+    public function run(Environment $environment) {
+        $linter = new TokenizerLinter();
+        return $linter->lintSource($this->source);
```
```suggestion
        return $linter->lintSource($this->source)->check();
```

staabm commented on src/Linter/PooledLintingResult.php:
```diff
@@ -0,0 +1,57 @@
+    private $promise;
+
+    /**
+     * @param Promise<TokenizerLintingResult> $promise
```
```suggestion
     * @param Promise $promise
```

staabm commented on src/Linter/PooledLintingResult.php:
```diff
@@ -0,0 +1,57 @@
+                $this->isSuccessful = $result->check();
+        }
+
+        return $this->isSuccessful;
```
```suggestion
```

staabm commented on src/Linter/PooledLintingResult.php:
```diff
@@ -0,0 +1,57 @@
+                /**
+                 * @var TokenizerLintingResult
+                 */
+                $result = \Amp\Promise\wait($this->promise);
```
```suggestion
                // will throw on error
                \Amp\Promise\wait($this->promise);
```

staabm commented on src/Linter/PooledLintingResult.php:
```diff
@@ -0,0 +1,57 @@
+                 * @var TokenizerLintingResult
+                 */
+                $result = \Amp\Promise\wait($this->promise);
+                $this->isSuccessful = $result->check();
```
```suggestion
```

staabm commented on src/Linter/PooledLintingResult.php:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function check()
```
check hat kein return

staabm commented on src/Linter/PooledLintingResult.php:
```diff
@@ -0,0 +1,57 @@
+    {
+        if (null === $this->isSuccessful) {
+                /**
+                 * @var TokenizerLintingResult
```
```suggestion
                 * @var bool
```

staabm commented on composer.json:
```diff
@@ -14,9 +14,10 @@
+        "php": "^7.0",
         "ext-json": "*",
         "ext-tokenizer": "*",
+        "amphp/parallel": "^1.2",
```
Update to 1.3

staabm commented:
As far as I can tell it seems to work now. After I changed the tasks to no longer throw Exception accross process boundaries it works.

actually it got not faster on my 2 core workstation.. need to test again on a machine with more cpu cores.

with this patch:
```
$ php php-cs-fixer fix ../redaxo/redaxo/src/core/
Loaded config default from "C:\xampp7.1.4\htdocs\PHP-CS-Fixer\.php_cs.dist".
Paths from configuration file have been overridden by paths provided as command
arguments.
string(30) "PhpCsFixer\Linter\PooledLinter"

Fixed all files in 32.071 seconds, 16.000 MB memory used

Files that were not fixed due to errors reported during linting before fixing:
   1) C:\xampp7.1.4\htdocs\redaxo\redaxo\src\core\lib\list.php

```

without this patch:
```
$ php php-cs-fixer fix ../redaxo/redaxo/src/core/
Loaded config default from "C:\xampp7.1.4\htdocs\PHP-CS-Fixer\.php_cs.dist".
Paths from configuration file have been overridden by paths provided as command
arguments.
string(33) "PhpCsFixer\Linter\TokenizerLinter"

Fixed all files in 29.905 seconds, 14.000 MB memory used

Files that were not fixed due to errors reported during linting before fixing:
   1) C:\xampp7.1.4\htdocs\redaxo\redaxo\src\core\lib\list.php

```

keradus commented:
@staabm , you won't see an improvement when parallelizing TokenizerLinter ( ref https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4838/files#r386713529 )

you can see improvement If you would parallelize fixing (one process/thread fixing one file with all rules, other process/thread fixing another files)

staabm commented:
> you won't see an improvement when parallelizing TokenizerLinter ( ref [#4838 (files)](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4838/files#r386713529)

the comment referenced does not mention why parallel linting cannot be faster?

> you can see improvement If you would parallelize fixing (one process/thread fixing one file with all rules, other process/thread fixing another files)

I tried this before.. the main problem we face is, that we cannot use Exception objects in IPC messages. that means that we would need to do the exception handling in the workers and concurrently write into the cacheManager (or alternatively use a separate "error-object" which supports IPC but would then dont include a proper stacktrace etc).

staabm commented:
Superceeded by https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5390
